### ISSUE TITLE
file write error: Input/output error with imuxsock module

### ISSUE THREAD

**Issue by macrotex** [rsyslog/rsyslog#4126]
[20-templates.conf.txt](https://github.com/rsyslog/rsyslog/files/4071994/20-templates.conf.txt)

### Expected behavior
rsyslog should log to syslog using the imuxsock module

### Actual behavior
rsyslog generates error messages like
`
file '9' write error: Input/output error [v8.1901.0 try https://www.rsyslog.com/e/2027 ]`

and does not write to the syslog.

### Steps to reproduce the behavior

Reboot machine and the service fails. If after rebooting the service is restarted the service starts working. It seems to fail to work only after a reboot. After reboot the journalctl logs has this:

```
Jan 15 11:35:27 radius02.example.com systemd[1]: Starting System Logging Service...
Jan 15 11:35:27 radius02.example.com rsyslogd[359]: imuxsock: Acquired UNIX socket '/run/systemd/journal/syslog' (fd 3) from systemd.  [v8.1901.0]
Jan 15 11:35:27 radius02.example.com systemd[1]: Started System Logging Service.
Jan 15 11:35:27 radius02.example.com rsyslogd[359]:  [origin software="rsyslogd" swVersion="8.1901.0" x-pid="359" x-info="https://www.rsyslog.com"] start
Jan 15 11:35:27 radius02.example.com rsyslogd[359]: file '9' write error: Input/output error [v8.1901.0 try https://www.rsyslog.com/e/2027 ]
Jan 15 11:35:27 radius02.example.com rsyslogd[359]: file '9' write error: Input/output error [v8.1901.0 try https://www.rsyslog.com/e/2027 ]
Jan 15 11:35:27 radius02.example.com rsyslogd[359]: file '9' write error: Input/output error [v8.1901.0 try https://www.rsyslog.com/e/2027 ]
...forever...
```

### Environment
- rsyslog version: 8.1901.0
- platform: Debian buster

Looking at the unix sockets, on a server where rsyslog is working properly I see these file descriptors owned by the rsyslogd process:
```
lr-x------ 1 root root 64 Jan 15 11:35 0 -> /dev/null
l-wx------ 1 root root 64 Jan 15 11:35 1 -> /dev/null
l-wx------ 1 root root 64 Jan 15 11:35 2 -> /dev/null
lrwx------ 1 root root 64 Jan 15 11:35 3 -> 'socket:[10211]'  (/run/systemd/journal/syslog)
lr-x------ 1 root root 64 Jan 15 11:35 4 -> /dev/urandom
lrwx------ 1 root root 64 Jan 15 11:35 5 -> 'socket:[14849]'  (/var/spool/postfix/dev/log)
lr-x------ 1 root root 64 Jan 15 11:35 6 -> /proc/kmsg
l-wx------ 1 root root 64 Jan 15 11:35 7 -> /dev/console
l-wx------ 1 root root 64 Jan 15 11:35 8 -> /var/log/messages
lrwx------ 1 root root 64 Jan 15 11:35 9 -> 'socket:[30511]'
```

On a server where rsyslog is _not_ working I see the same set _except_ the file descriptor corresponding to the number in the error message (in the example above this is '9') is missing and the last socket (the one not pointing at a file) keeps getting re-generated.

Rsyslog is managed by systemd with the stock systemd unit file that comes with Debian buster.

```
# /etc/rsyslog.conf
$ModLoad imuxsock
$ModLoad imklog
$ActionFileDefaultTemplate RSYSLOG_TraditionalFileFormat
$FileOwner root
$FileGroup root
$WorkDirectory /var/spool/rsyslog
$FileCreateMode 0640
$DirCreateMode 0755
$SystemLogRateLimitInterval 0
$IncludeConfig /etc/rsyslog.d/*.conf

# /etc/rsyslog.d/postfix.conf
$AddUnixListenSocket /var/spool/postfix/dev/log

# /etc/rsyslog.d/95-default.conf
*.emerg action(type = "omusrmsg"
               name  = "emerg"
               users = "root")
*.debug action(type = "omfile"
               name     = "defaultLocal"
               file     = "/var/log/messages"
               template = "FileFormat")
*.err   action(type = "omfile"
               name = "err"
               file = "/dev/console")

# /etc/rsyslog.d/50-authproxy.conf
if $syslogfacility-text == 'local3' then {
   action(type     = "omfile"
          name     = "authproxyLocal"
          file     = "/var/log/duoauthproxy"
          template = "FileFormat")
   stop
}

#  /etc/rsyslog.d/20-templates.conf
# Attached as 20-templates.conf.txt
```
Here is the systemd unit file used to control the rsyslogd service on buster:
```
[Unit]
Description=System Logging Service
Requires=syslog.socket
Documentation=man:rsyslogd(8)
Documentation=https://www.rsyslog.com/doc/

[Service]
Type=notify
ExecStart=/usr/sbin/rsyslogd -n -iNONE
StandardOutput=null
Restart=on-failure

# Increase the default a bit in order to allow many simultaneous
# files to be monitored, we might need a lot of fds.
LimitNOFILE=16384

[Install]
WantedBy=multi-user.target
Alias=syslog.service
```

**Comment by davidelang** [comment]
are you saying that the stock syslog configuration is opening 
/var/spool/postfix/dev/log and whatever this additional socket on FD 9 is?

Please provide the full config as we are not familiar with what Debian Buster 
ships by default? This includes anything that is included (the contents of 
/etc/rsyslog.d/*.conf)

David Lang

**Comment by macrotex** [comment]
> 
> are you saying that the stock syslog configuration is opening /var/spool/postfix/dev/log and whatever this additional socket on FD 9 is? Please provide the full config as we are not familiar with what Debian Buster ships by default? This includes anything that is included (the contents of /etc/rsyslog.d/*.conf) David Lang

I have updated the original description to include all the configuration files as well as the systemd unit file used to control the service.

The reason I mention the sockets is that a symptom I am seeing on the servers where rsyslogd is not running properly is an error mentioning a reference to a file descriptor which, I believe, is supposed to be a socket that has vanished. My guess is that rsyslogd starts up, creates a socket with file descriptor N, this socket goes away for some unknown reason, but rsyslogd keeps looking for that now-missing socket. 

On a server with the problem I see millions of these errors in the systemd journal:
`Jan 16 07:25:14 radius02.stanford.edu rsyslogd[12305]: file '7' write error: Input/output error [v8.1901.0 try https://www.rsyslog.com/e/2027 ]`

When I list the file descriptors associated with the rsyslogd process on this server I see that file descriptor 7 is having a problem:
```
# ls -l /proc/12305/fd/
ls: cannot read symbolic link '/proc/12305/fd/7': No such file or directory
total 0
lr-x------ 1 root root 64 Jan 15 23:46 0 -> /dev/null
l-wx------ 1 root root 64 Jan 15 23:46 1 -> /dev/null
l-wx------ 1 root root 64 Jan 15 23:46 2 -> /dev/null
lrwx------ 1 root root 64 Jan 15 23:46 3 -> 'socket:[10211]'
lr-x------ 1 root root 64 Jan 15 23:46 4 -> /dev/urandom
lrwx------ 1 root root 64 Jan 15 23:46 5 -> 'socket:[137154]'
lr-x------ 1 root root 64 Jan 15 23:46 6 -> /proc/kmsg
l-wx------ 1 root root 64 Jan 15 23:46 7
l-wx------ 1 root root 64 Jan 15 23:46 8 -> /var/log/messages
lrwx------ 1 root root 64 Jan 15 23:46 9 -> 'socket:[219645]'
```

**Comment by macrotex** [comment]
We have solved the issue. It turns out we had broken `/dev/console`, so on boot up when rsyslog tried to write to `/dev/console` it could not. After the system had started and we restarted rsyslog, since nothing was writing to `/dev/console` we did not see any errors. Fixing `/dev/console` fixed rsyslog.

This section of the rsyslog configuration tells rsyslog to send errors to `/dev/console`:
```
*.err   action(type = "omfile"
               name = "err"
               file = "/dev/console")
```

Perhaps rsyslog could give better feedback on which part of its configuration is generating the error?
